File: typical90/python/011.py
def main():
    N = int(input())

    table = []
    for _ in range(N):
        # shape: (D, C, S)
        table.append(tuple(map(int, input().split())))

    # Dの昇順でソート
    table.sort(key=lambda x: x[0])
    # j + C[i + 1]でおおきくなりがちなので上限まで容量確保しておくほうが良さそう
    D = [0] * 5009
    C = [0] * 5009
    S = [0] * 5009
    for i in range(N):
        D[i] = table[i][0]
        C[i] = table[i][1]
        S[i] = table[i][2]

    dp = [[0] * 5009 for _ in range(5009)]
    # 動的計画法 O(N^2)
    for i in range(N):
        for j in range(5000):
            # dp[i + 1][j]ははじめ0でinitiallizeされてるのでmaxをとって上書きしてく
            # 仕事iをやらない場合
            dp[i + 1][j] = max(dp[i + 1][j], dp[i][j])
            # 仕事iをやる場合
            # 締切に間に合ってるか
            # j日目+仕事iにかかる日数C_i <= D_i
            if j + C[i] <= D[i]:
                dp[i + 1][j + C[i]] = max(
                    dp[i + 1][j + C[i]], dp[i][j] + S[i]
                )

    ans = 0
    for j in range(5009):
        ans = max(ans, dp[N][j])

    print(ans)


if __name__ == "__main__":
    main()

File: typical90/rust/src/bin/011.rs
use proconio::{input, fastout};

#[allow(dead_code)]
fn solve1() {
    input! {
       n: usize,
       mut dcs: [(usize, usize, usize); n],
    }
    dcs.sort();

    let d_max = dcs.last().unwrap().0;
    let mut dp = vec![0; d_max + 1];
    // DP
    for (d, c, s) in dcs {
        for i in (0..=d_max).rev() {
            if c > i || d < i {
                continue;
            }
            dp[i] = dp[i].max(dp[i - c] + s);
        }
    }

    println!("{}", dp.iter().max().unwrap());
}

fn solve2() {
    input! {
       n: usize,
       mut dcs: [(usize, usize, usize); n],
    }
    // d(締切日)で昇順にソートする
    dcs.sort();
    
    let mut dp = vec![vec![0; 5009]; 5009];
    for i in 0..n - 1 {
        for j in 0..=5000 {
            let (d, c, s) = dcs[i + 1];
            // 仕事i+1をやらない場合
            dp[i + 1][j] = std::cmp::max(dp[i + 1][j], dp[i][j]);

            // 仕事i+1をやる場合
            if j + c <= d {
                dp[i + 1][j + c] = std::cmp::max(dp[i + 1][j + c], dp[i][j] + s);
            }
        }
    }
    let mut ans = 0;
    for i in 0..=5000 {
        ans = std::cmp::max(ans, dp[n - 1][i]);
    }
    println!("{}", ans);
}


#[fastout]
fn main() {
    solve2()
}
